### PR TITLE
GPIO: do not overwrite interrupt handler

### LINKED ISSUE #2009: The `Io` struct has multiple constructors which makes it confusing to use correctly
As a workaround, in #1861 the `Io::new_no_bind_interrupt` constructor was added to allow initializing the `Io` struct _without_ binding the `GPIO` interrupt.

This is a bit of a footgun, and is too easy to mess up. We should try to find a better way to handle this scenario.

### PR DESCRIPTION
**Pull request by bugadani**
Another idea I'd want some feedback on. Closes #2009 I hope.

Why not keep this in `Io::new`? I intend to move the Pins out into Peripherals so that users don't need to create `Io` - unless they want to handle user interrupts. That is a bigger change, though.

### CODE REVIEW

**Comment by jessebraham** [comment]
I'm not sure I understand these changes, it seems to solve one problem while creating another. As far as I can tell, if you do not want to bind the interrupt handler you can no longer call `esp_hal::init()`? This will result in a large amount of boilerplate required for initializing the WDTs, clocks, etc. in this case.

**Comment by jessebraham** [comment]
> In which case do you explicitly not want to bind any interrupt handlers?

When using e.g. RTIC and using hardware tasks involving GPIO. This was the reason the `Io::new_no_bind_interrupt()` constructor was added in the first place. I've confirmed that using this branch and calling `esp_hal::init()` causes the `sw_and_hw` example to stop functioning as expected.

**Review by jessebraham — APPROVED** [review]
Can confirm things are working as expected again, thanks for taking care of that!

File: hil-test/tests/gpio_custom_handler.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        io.set_interrupt_handler(interrupt_handler);
+
+        let (gpio1, gpio2) = hil_test::common_test_pins!(io);
+
+        let timg0 = TimerGroup::new(peripherals.TIMG0);
+        esp_hal_embassy::init(timg0.timer0);
+
+        let counter = drive_pins(gpio1, gpio2).await;
+
+        // GPIO is bound to something else, so we don't expect the async API to work.
```
Comment (bjoernQ) [inline hil-test/tests/gpio_custom_handler.rs:96]: that comment needs to be changed

**Review by bjoernQ — APPROVED** [review]
LGTM

**Pull request merged**